File: src/main/java/frc/robot/io/TalonFXMotorIOSim.java
package frc.robot.io;

import com.ctre.phoenix6.configs.CANcoderConfiguration;
import com.ctre.phoenix6.configs.TalonFXConfiguration;
import com.ctre.phoenix6.controls.Follower;
import com.ctre.phoenix6.sim.TalonFXSimState;
import edu.wpi.first.math.system.plant.DCMotor;
import java.util.Optional;

/** Simulation-backed Talon FX IO wrapper. */
public class TalonFXMotorIOSim extends TalonFXMotorIO {
  private static final String KRAKEN_X60 = "KrakenX60";
  private static final String KRAKEN_X60_FOC = "KrakenX60_FOC";
  private static final String KRAKEN_X44 = "KrakenX44";
  private static final String KRAKEN_X44_FOC = "KrakenX44_FOC";

  private final DCMotor motorModel;

  /** Creates a simulated Talon FX with no follower or CANcoder. */
  public TalonFXMotorIOSim(
      int canID, TalonFXConfiguration config, String motorType, String motorName) {
    this(canID, config, motorType, motorName, Optional.empty(), Optional.empty());
  }

  /** Creates a simulated Talon FX configured as a follower. */
  public TalonFXMotorIOSim(
      int canID,
      TalonFXConfiguration config,
      String motorType,
      String motorName,
      Follower followerConfig) {
    this(
        canID, config, motorType, motorName, Optional.empty(), Optional.ofNullable(followerConfig));
  }

  /** Creates a simulated Talon FX using external CANcoder feedback. */
  public TalonFXMotorIOSim(
      int canID,
      TalonFXConfiguration config,
      String motorType,
      String motorName,
      CANcoderConfiguration canCoderConfig) {
    this(
        canID, config, motorType, motorName, Optional.ofNullable(canCoderConfig), Optional.empty());
  }

  /** Full constructor for selecting optional CANcoder/follower config and motor model. */
  public TalonFXMotorIOSim(
      int canID,
      TalonFXConfiguration config,
      String motorType,
      String motorName,
      Optional<CANcoderConfiguration> canCoderConfig,
      Optional<Follower> followerConfig) {
    super(canID, config, motorName, canCoderConfig, followerConfig);
    this.motorModel = resolveMotorType(motorType);
  }

  /** Returns the Phoenix simulation state for this Talon FX. */
  public TalonFXSimState getSimState() {
    return motor.getSimState();
  }

  /** Returns the WPILib motor model used by mechanism simulations. */
  public DCMotor getMotorType() {
    return motorModel;
  }

  private static DCMotor resolveMotorType(String motorType) {
    return switch (motorType) {
      case KRAKEN_X60 -> DCMotor.getKrakenX60(1);
      case KRAKEN_X60_FOC -> DCMotor.getKrakenX60Foc(1);
      case KRAKEN_X44 -> DCMotor.getKrakenX44(1);
      case KRAKEN_X44_FOC -> DCMotor.getKrakenX44Foc(1);
      default -> DCMotor.getKrakenX60Foc(1);
    };
  }
}

File: src/main/cpp/io/TalonFXMotorIOSim.cpp
#include "io/TalonFXMotorIOSim.h"

using namespace ctre::phoenix6;

TalonFXMotorIOSim::TalonFXMotorIOSim(
    int canID, configs::TalonFXConfiguration config,
    const std::string& motorType, std::string motorName)
    : TalonFXMotorIOSim(canID, std::move(config), motorType, std::move(motorName),
                        std::nullopt, std::nullopt) {}

TalonFXMotorIOSim::TalonFXMotorIOSim(
    int canID, configs::TalonFXConfiguration config,
    const std::string& motorType, std::string motorName,
    controls::Follower followerConfig)
    : TalonFXMotorIOSim(canID, std::move(config), motorType, std::move(motorName),
                        std::nullopt, std::make_optional(std::move(followerConfig))) {}

TalonFXMotorIOSim::TalonFXMotorIOSim(
    int canID, configs::TalonFXConfiguration config,
    const std::string& motorType, std::string motorName,
    configs::CANcoderConfiguration canCoderConfig)
    : TalonFXMotorIOSim(canID, std::move(config), motorType, std::move(motorName),
                        std::make_optional(std::move(canCoderConfig)), std::nullopt) {}

TalonFXMotorIOSim::TalonFXMotorIOSim(
    int canID, configs::TalonFXConfiguration config,
    const std::string& motorType, std::string motorName,
    std::optional<configs::CANcoderConfiguration> canCoderConfig,
    std::optional<controls::Follower> followerConfig)
    : TalonFXMotorIO(canID, std::move(config), std::move(motorName),
                     std::move(canCoderConfig), std::move(followerConfig)),
      m_motorModel{ResolveMotorType(motorType)} {}

sim::TalonFXSimState& TalonFXMotorIOSim::GetSimState() {
  return m_motor.GetSimState();
}

frc::DCMotor TalonFXMotorIOSim::ResolveMotorType(const std::string& motorType) {
  if (motorType == "KrakenX60") return frc::DCMotor::KrakenX60(1);
  if (motorType == "KrakenX60_FOC") return frc::DCMotor::KrakenX60FOC(1);
  if (motorType == "KrakenX44") return frc::DCMotor::KrakenX44(1);
  if (motorType == "KrakenX44_FOC") return frc::DCMotor::KrakenX44FOC(1);
  return frc::DCMotor::KrakenX60FOC(1);
}
